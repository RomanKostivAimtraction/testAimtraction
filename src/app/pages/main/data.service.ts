import { Injectable } from '@angular/core';
import { ImageComponent } from '../main/image/image.component';


@Injectable({
  providedIn: 'root'
})
export class DataService {

  columnDefs = [
    {
      headerName: '',
      field: 'checkBox',
      checkboxSelection: true,
      headerCheckboxSelection: true,
      width: 10
    },
    {
      headerName: '',
      field: 'thumbnails',
      sortable: true,
      filter: true,
      cellRendererFramework: ImageComponent
    },
    {
      headerName: 'Published on',
      field: 'publishedAt',
      sortable: true,
      filter: true,
      valueFormatter: params => this.dataFormater(params.data.publishedAt)
    },
    {
      headerName: 'Video Title',
      field: 'title',
      sortable: true,
      filter: true,
      // preventDefaultOnContextMenu: onCellContextMenu,
      cellRenderer: params => this.videoTitleLink(params),
    },
    { headerName: 'Description', field: 'description', sortable: true, filter: true, },
  ];
  constructor() { }

  dataFormater(date) {
    return `${date.slice(8, 10)}.${date.slice(5, 7)}.${date.slice(0, 4)}`;
  }


  videoTitleLink(params) {
    const [id, title] = [params.value.id.videoId, params.value.snippet.title];
    return `<a href="https://www.youtube.com/watch?v=${id}"target="_blank" rel="noopener">${title}</a>`;
  }
}
