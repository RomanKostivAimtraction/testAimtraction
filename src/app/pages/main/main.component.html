<form [formGroup]='myForm'>
    <div class="row mt-4 pl-4">

        <div class="col-md-3">
            <mat-form-field>
                <input matInput type="text" placeholder="Search..." formControlName='searchValue'>
            </mat-form-field>
        </div>
        <div class="col-md-1">
            <mat-icon matDatepickerToggleIcon svgIcon="search" (click)="getInfo()" matTooltip="press to search"
                [matTooltipPosition]="'above'">
            </mat-icon>
            <!-- <button  class="buttonClick">Search</button> -->
        </div>
        <div class="col-md-1">
            <mat-form-field>
                <mat-label>All rows</mat-label>
                <input matInput type="number" [value]='countAllRows$ | async'>
            </mat-form-field>
        </div>
        <div class="col-md-1">
            <mat-form-field>
                <mat-label>Seected rows</mat-label>
                <input matInput type="number" [value]='countSelectedRows$ | async'>
            </mat-form-field>
        </div>
        <div class="col-md-1">

            <mat-slide-toggle formControlName='mode' (change)="changeMode($event.checked)">Slide me!</mat-slide-toggle>

        </div>

    </div>
</form>
<div class="row">
    <div class="spiner" *ngIf="spinerIsLoading">
        <img src="assets/icons/spiner.gif" alt="">
    </div>
</div>




<ag-grid-angular style="width: 100%; height: 100%;" class="ag-theme-alpine" [gridOptions]="gridOptions"
    [rowData]="rowData" [modules]="modules" [columnDefs]="dataService.columnDefs" rowSelection='multiple'
    (rowSelected)="rowSelected()" [getContextMenuItems]="getContextMenuItems">
</ag-grid-angular>