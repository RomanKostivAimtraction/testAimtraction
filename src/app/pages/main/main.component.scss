.panel {
    width: 100%;
    height: 50px;
    display: flex;
    // flex-wrap: nowrap;
    margin-bottom: 50px;
    input {
        height: 40px;
        // font-weight: bold;
        font-size: 20px;
        padding-left: 10px;
        // padding: 5px;
        color: grey;
        border: 1px solid grey;
        border-radius: 5px;
    }
    .searchPlace {
        .buttonClick {
            margin-left: 30px;
            cursor: pointer;
        }
    }
    .infoPanel {
        width: 50%;
        height: 50px;
        display: flex;
        margin-left: 150px;
        justify-content: space-around;
        .box {
            display: flex;
            flex-direction: column;
            span{
               color: gray;
            }
        }
        input {
            text-align: center;
        }
    }

    .spiner {
        width: 50px;
        height: 50px;
        // margin-top: 30px;
        margin: 30px auto;
        img {
            width: 50px;
            height: 50px;
        }
    }
}
