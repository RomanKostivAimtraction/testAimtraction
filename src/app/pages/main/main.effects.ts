import { Injectable } from "@angular/core";
import { Actions, createEffect } from "@ngrx/effects";

@Injectable()

export class mainEffects {
    constructor(
        private actions$: Actions
    ){}

    
}