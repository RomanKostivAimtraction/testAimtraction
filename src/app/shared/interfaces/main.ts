export interface objFromYoutube {
    // snippet?: {
    //     thumbnails: {
    //         medium: {
    //             url: string
    //         }
    //     },
    //     publishedAt: string,
    //     description: string
    // }
}

export interface IData {
    // items?: Array<{ snippet: { thumbnails: { medium: { url: string } }, publishedAt: string, description: string } }>;
    items: IRowData[]
    countAllRows?: number;
    countSelectedRows?: number;
    error?: any;
    isLoading?: boolean;
}

export interface IRowData {
    description: string;
    publishedAt: string;
    thumbnails: string;
    title: object;
}