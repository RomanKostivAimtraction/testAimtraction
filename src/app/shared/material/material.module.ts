import { NgModule } from '@angular/core';
import { MatToolbarModule } from '@angular/material/toolbar';
import { MatFormFieldModule } from '@angular/material/form-field';
import { MatInputModule } from '@angular/material/input';
import { MatSlideToggleModule } from '@angular/material/slide-toggle';
import { MatIconModule } from '@angular/material/icon';
import { MatTooltipModule } from '@angular/material/tooltip';

const MaterialComponents = [
  MatToolbarModule,
  MatFormFieldModule,
  MatInputModule,
  MatSlideToggleModule,
  MatIconModule,
  MatTooltipModule

];

@NgModule({
  declarations: [],
  imports: [MaterialComponents],
  exports: [MaterialComponents],
})
export class MaterialModule { }
