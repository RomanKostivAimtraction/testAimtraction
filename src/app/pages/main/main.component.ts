import { GridOptions, Module } from '@ag-grid-community/core';
import { Component, OnDestroy, OnInit } from '@angular/core';
import { RequestsService } from '../../shared/services/requests.service';
import { DataService } from './data.service';
import { take } from 'rxjs/operators';

import { ClientSideRowModelModule } from '@ag-grid-community/client-side-row-model';
import { Store } from '@ngrx/store';
import { countAllRowsSelector, countAllRows, countSelectedRows, countSelectedRowsSelector } from 'src/app/reducers/main.reducer';
// import { Clipboard } from "@angular/cdk/clipboard"

interface IRowData {
  description: string;
  publishedAt: string;
  thumbnails: string;
  title: object;
}

interface IDataYoutube {
  items?: Array<{ snippet: { thumbnails: { medium: { url: string } }, publishedAt: string, description: string } }>;
  etag?: string;
  kind?: string;
  nextPageToken?: string;
  regionCode?: string;
  pageInfo?: object;
}

// interface IContextMenu {
//   name: string,
//   openNewTab: Function,
//   cssClasses: Array<string>,
// }


@Component({
  selector: 'app-main',
  templateUrl: './main.component.html',
  styleUrls: ['./main.component.scss'],
})
export class MainComponent implements OnInit, OnDestroy {
  searchValue = '';
  // countAllcase: number;
  // countSelectedCase: number;
  public gridOptions: GridOptions;
  modules: Module[] = [ClientSideRowModelModule];

  spinerIsLoading = false;

  rowData: IRowData[];

  getContextMenuItems: any;

  request: any;

  // changeMode$ = this.store.select(modeChecboxSelector);
  countAllRows$ = this.store.select(countAllRowsSelector);
  countSelectedRows$ = this.store.select(countSelectedRowsSelector);




  constructor(
    private requestService: RequestsService,
    public dataService: DataService,
    private store: Store
    // private clipboard: Clipboard
  ) {
    this.gridOptions = {
      onGridReady: () => {
        this.gridOptions.api.sizeColumnsToFit();
      },
      rowHeight: 200,
      // copyHeadersToClipboard: true
    };
  }

  ngOnInit(): void {
    this.getContextMenuItems = this.getContextMenuItemsFunc;
  }

  getContextMenuItemsFunc(params) {
    console.log(params);
    const contextArray: Array<any> = [ // need know true interfase for array
      'copy',
      'copyWithHeaders'
    ];

    if (params.column.colDef.field === 'title') {
      contextArray.push('separator');
      contextArray.push({
        name: 'Open in new tab',
        action() {
          window.open(`https://www.youtube.com/watch?v=${params.node.data.title.id.videoId}`, '_blank');
        },
        cssClasses: ['redFont', 'bold'],
      });

    }

    return contextArray;
  }


  rowSelected() {
    console.log(this.gridOptions.api.getSelectedRows());
    const countRow = this.gridOptions.api.getSelectedRows().length;
    // this.countSelectedCase = this.gridOptions.api.getSelectedRows().length;
    this.store.dispatch(countSelectedRows());
  }

  changeMode() {
    this.dataService.columnDefs[0].checkboxSelection = !this.dataService.columnDefs[0].checkboxSelection;
    this.dataService.columnDefs[0].headerCheckboxSelection = !this.dataService.columnDefs[0].headerCheckboxSelection;
    this.gridOptions.api.setColumnDefs(this.dataService.columnDefs);
    this.gridOptions.api.refreshCells();
  }

  getInfo() {

    this.spinerIsLoading = true;
    const apiKey = 'AIzaSyB3GVWc8NIjn8B2-BbzW-AOko2lfOHgTKw';
    const url = `https://www.googleapis.com/youtube/v3/search?key=${apiKey}&maxResults=20&type=video&part=snippet&q=${this.searchValue}`;
    this.request = this.requestService.getInfo<IDataYoutube>(url).pipe(take(1)).subscribe(res => {
      console.log('answare from youtube');
      console.log(res);

      this.spinerIsLoading = false;
      // this.countAllcase = res.items.length;
      this.store.dispatch(countArrRows());

      this.rowData = res.items.map(elem => {
        return {
          thumbnails: elem.snippet.thumbnails.medium.url,
          publishedAt: elem.snippet.publishedAt,
          title: elem,
          description: elem.snippet.description,
        };
      });
      console.log('rowData');
      console.log(this.rowData);
    }, (err: any) => { });
  }




  ngOnDestroy(): void {
    // this.request.unsubscribe();
  }

}
