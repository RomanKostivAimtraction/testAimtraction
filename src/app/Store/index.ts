import {
  ActionReducer,
  ActionReducerMap,
  createFeatureSelector,
  createSelector,
  MetaReducer
} from '@ngrx/store';
import { environment } from '../../environments/environment';
import { IData } from '../shared/interfaces/main';
import { mainReducer, SELECTROWS_KEY } from './main.reducer';

export interface State {
  [SELECTROWS_KEY]: IData;

}

export const reducers: ActionReducerMap<State> = {
  [SELECTROWS_KEY]: mainReducer,

};


export const metaReducers: MetaReducer<State>[] = !environment.production ? [] : [];
