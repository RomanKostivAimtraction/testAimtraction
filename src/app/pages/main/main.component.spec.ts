import { AgGridModule } from '@ag-grid-community/angular';
import { HttpClientModule } from '@angular/common/http';
import { ComponentFixture, TestBed, waitForAsync } from '@angular/core/testing';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { StoreModule } from '@ngrx/store';
import { MaterialModule } from '../../shared/material/material.module';
import { ToastrModule, ToastrService } from 'ngx-toastr';
import { provideMockStore, MockStore } from '@ngrx/store/testing';
import { countAllRowsSelector, countSelectedRows, countSelectedRowsSelector, dataState, requeftFromYoutube, spinerIsLoadingSelector, YTdataSelector } from 'src/app/Store/main.reducer';
import * as mainReducer from '../../Store/main.reducer';



import { MainComponent } from './main.component';
import { reducers, metaReducers } from '../../Store';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
// import { SET_test } from './data.service';
import { IParamsForContextMenu } from 'src/app/shared/interfaces/main';
import { MenuItemDef } from '@ag-grid-community/core';
import { Subject } from 'rxjs';

declare const expect: jest.Expect;

describe('MainComponent', () => {
  let component: MainComponent;
  let fixture: ComponentFixture<MainComponent>;
  let store: MockStore;
  let toastr: ToastrService

  const initialState = {
    data: mainReducer.dataState
  };

  beforeEach(waitForAsync(() => {
    TestBed.configureTestingModule({
      declarations: [MainComponent],
      imports: [
        FormsModule,
        ReactiveFormsModule,
        MaterialModule,
        BrowserAnimationsModule,
        ToastrModule.forRoot(),
        StoreModule.forRoot(reducers, {
          metaReducers
        }),

        AgGridModule.withComponents([MainComponent]),
        HttpClientModule
      ],
      providers: [ToastrService, provideMockStore({ initialState })],
    })
      .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(MainComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
    store = TestBed.inject(MockStore);
    toastr = TestBed.inject(ToastrService)
  });


  it('should mark form as touched if form invalid', () => {
    const spyForm = jest.spyOn(component.myForm, 'markAllAsTouched');
    const spyToastr = jest.spyOn(toastr, 'warning');
    component.getInfo();
    expect(spyForm).toHaveBeenCalled();
    expect(spyToastr).toHaveBeenCalled();
  })

  it('should mark form as touched if form uninvalid', () => {
    const spyStoreDispatch = jest.spyOn(store, 'dispatch');
    component.myForm.get('searchValue').setValue('marvel')
    const searchValue = component.myForm.get('searchValue').value
    component.getInfo();
    expect(spyStoreDispatch).toHaveBeenCalledWith(requeftFromYoutube({ searchValue: searchValue}));
  })



  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should dispatch count row in store', () => {
    const spyStoreDispatch = jest.spyOn(store, 'dispatch');
    const countRow: number = 0;
    component.rowSelected();
    expect(spyStoreDispatch).toHaveBeenCalledWith(countSelectedRows({ countRow }));
  })


  it('should return context menu with open new tab', () => {
    const params = {
      column: { colDef: { field: 'title' } },
      node: { data: { title: { id: { videoId: 'df4g5hgrh' } } } }
    }
    const contextMenuResult = component.getContextMenuItemsFunc(params as IParamsForContextMenu);
    expect(contextMenuResult).toMatchSnapshot();
  });

  it('should return context menu without open new tab', () => {
    const params = {
      column: { colDef: { field: 'dsfdfg' } },
      node: { data: { title: { id: { videoId: 'df4g5hgrh' } } } }
    }
    const contextMenuResult = component.getContextMenuItemsFunc(params as IParamsForContextMenu);
    const contextMenu: (string | MenuItemDef)[] = ['copy', 'copyWithHeaders'];
    expect(contextMenuResult).toEqual(contextMenu);
  });



  it('should be change mode ON', () => {
    const value = true;
    const spyDeselectAll = jest.spyOn(component.gridOptions.api, 'deselectAll')
    component.changeMode(value);
    expect(spyDeselectAll).toHaveBeenCalled();
  })

  it('should be change mode OFF', () => {
    const value = false;
    const spySetColumnDefs = jest.spyOn(component.gridOptions.api, 'setColumnDefs')
    const spyRefreshCells = jest.spyOn(component.gridOptions.api, 'refreshCells')
    component.changeMode(value);
    expect(component.dataService.columnDefs[0].checkboxSelection).toEqual(value);
    expect(component.dataService.columnDefs[0].headerCheckboxSelection).toEqual(value);
    expect(spySetColumnDefs).toHaveBeenCalledWith(component.dataService.columnDefs);
    expect(spyRefreshCells).toHaveBeenCalled();
  })

  it('should be contain reactive form', () => {
    component.reactiveForm();
    const formValue = { searchValue: '', mode: true }
    const result = component.myForm.value;
    expect(result).toEqual(formValue)
  })


  it('shold be undiscrubed', () => {
    const spy = jest.spyOn(component.destroy$, "next");
    const spy1 = jest.spyOn(component.destroy$, "complete");
    component.ngOnDestroy();
    expect(spy).toHaveBeenCalled();
    expect(spy1).toHaveBeenCalled();
  })
});


