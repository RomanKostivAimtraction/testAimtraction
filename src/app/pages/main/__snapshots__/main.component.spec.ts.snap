// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainComponent should return context menu with open new tab 1`] = `
Array [
  "copy",
  "copyWithHeaders",
  "separator",
  Object {
    "action": [Function],
    "cssClasses": Array [
      "redFont",
      "bold",
    ],
    "name": "Open in new tab",
  },
]
`;
